docs_src/design/adr/system-management/0006-EdgeX-Metrics.md

```diff
@@ -0,0 +1,115 @@
+* Have services send their metrics objects in Event/Reading form through Core Data where (special) application services could filter out and deal with metrics data.  Under this design then:
+  * Metadata descriptions along with typing/tagging value descriptors (or their equivalent) would define control plane data from each service.  In terms of data, control plane data would look no different than sensor data.  The difference is that the data could come from every service (versus device services).
+  * Special tags or marks in the current Event/reading model may be required to designated metric data versus standard sensor data.  The beauty of this design is that it requires no or minimal changes to services to implement (although each service must add code to collect control plane metrics and send them to core data.
+  * Application services would filter, transform, and otherwise export control plane data (as it does sensor data today) in the same fashion as it gets sensor data to 3rd parties.  It would allow rules engine to fire on the part of control plane data – example command a device or eventually send an alert/notification.  A special application service could be used to deal with particulars of control plane export if necessary.
+  * The existing system management service could would not have to change.  It would still provide for start/top/restart operations, service level metrics (CPU, memory), and provide get/set configuration operations.  If required, the SMA could also provide control plane data queries for a request/response style interface for control plane data, while the application services provide a push of control plane data to HTTP or even message bus endpoints.
+* A new service could be created to serve as the collection point, and allow that service (and lots of alternate implementations of that service) to do what it wishes with the metrics data on arrival (persist it, store and forward, forward it via message bus, etc.).  We could offer some reference implementations or simply have an implementation that logs the metrics data just to be a placeholder and implementation of any API on the service.
+* Do nothing special - simply allow each service to send its metrics data to a specified message endpoint configured in the service.  This could even be different message endpoints for each service.  This would provide ultimate flexibility but also put the onus on the end user to architect a solution on collection and movement of metric data.
+
+#### Integration of 3rd party tools ####
+While EdgeX could provide integration to something like Prometheus, it is viewed that this exercise is best accomplished by end users or commercialization efforts.  Metrics collection should help facilitate use of tool like Prometheus and make it easier to integrate (perhaps even offering example code in holding), but EdgeX should avoid any tight integration to specific tools.
```

Comment:
I think the requirement list very closely mirrors that available in prometheus. In the spirit of not-invented-here / don't reinvent the wheel I think it would be valuable to seriously how closely prometheus meets the requirements. I like the approach of make metric collection but offer the 3rd party tool as a separate guide / sample 👍 

I'd strongly encourage adopting protocols / standards that allow for easy plugability of a 3rd party tool:
https://opentelemetry.io/ _would offer lightspeed, zipkin, jaeger and datadog compatibility to name a few_
https://github.com/statsd/ _statsd would offer prometheus and datadog to name a few_
In both cases there's fairly decent golang support too

Kong has prometheus, zipkin and statsd support and happens to have easy access to a number of the aforementioned metrics.

https://docs.konghq.com/hub/kong-inc/prometheus/
https://docs.konghq.com/hub/kong-inc/zipkin/
https://docs.konghq.com/hub/kong-inc/statsd/